libr/bin/format/wasm/wasm.c

```diff
@@ -266,527 +243,368 @@ static RList *r_bin_wasm_get_type_entries(RBinWasmObj *bin, RBinWasmSection *sec
 	}
-	if (!sec) {
-		r_list_free (ret);
-		return NULL;
+	if (!(consume_u32_r (b, max, &ptr->body_size))) {
+		goto beach;
 	}
-	RBuffer *b = bin->buf;
-	r_buf_seek (b, sec->payload_data, R_IO_SEEK_SET);
-	ut64 max = r_buf_tell (b) + sec->payload_len - 1;
-	if (!(max < r_buf_size (b))) {
+	ut32 j = r_buf_tell (b);
+	if (!(r_buf_tell (b) + ptr->body_size - 1 <= max)) {
 		goto beach;
 	}
-	ut32 r = 0;
-	while (r_buf_tell (b) <= max && r < sec->count) {
-		if (!(ptr = R_NEW0 (RBinWasmDataEntry))) {
-			return ret;
-		}
-		if (!(consume_u32_r (b, max, &ptr->index))) {
-			goto beach;
-		}
-		if (!(ptr->offset.len = consume_init_expr_r (b, max, R_BIN_WASM_END_OF_CODE, NULL))) {
-			goto beach;
-		}
-		if (!(consume_u32_r (b, max, &ptr->size))) {
-			goto beach;
-		}
-		ptr->data = r_buf_tell (b);
-		r_buf_seek (b, ptr->size, R_IO_SEEK_CUR);
-		if (!r_list_append (ret, ptr)) {
-			free (ptr);
-			// should it jump to beach?
-		}
-		ptr = NULL;
-		r++;
+	if (!(consume_u32_r (b, max, &ptr->local_count))) {
+		goto beach;
 	}
-	return ret;
+	if (consume_locals_r (b, max, ptr) < ptr->local_count) {
+		goto beach;
+	}
+	ptr->code = r_buf_tell (b);
+	ptr->len = ptr->body_size - ptr->code + j;
+	r_buf_seek (b, ptr->len - 1, R_IO_SEEK_CUR); // consume bytecode
+	r_buf_read (b, &ptr->byte, 1);
+	if (ptr->byte != R_BIN_WASM_END_OF_CODE) {
+		goto beach;
+	}
+	return ptr;
+
 beach:
-	eprintf ("[wasm] error: beach data entries\n");
-	free (ptr);
-	return ret;
+	r_bin_wasm_free_codes (ptr);
+	return NULL;
 }
 
-static RList *r_bin_wasm_get_symtab_entries(RBinWasmObj *bin, RBinWasmSection *sec) {
-	RList *ret = NULL;
-	RBinWasmSymbol *ptr = NULL;
-	size_t read = 0;
-	if (!(ret = r_list_newf ((RListFree)free))) {
+static void *parse_data_entry(RBuffer *b, ut64 max) {
+	RBinWasmDataEntry *ptr = R_NEW0 (RBinWasmDataEntry);
+	if (!ptr) {
 		return NULL;
 	}
-	if (!sec) {
-		r_list_free (ret);
-		return NULL;
+	if (!(consume_u32_r (b, max, &ptr->index))) {
+		goto beach;
 	}
-	RBuffer *b = bin->buf;
-	r_buf_seek (b, sec->payload_data + 3, R_IO_SEEK_SET);
-	ut64 max = r_buf_tell (b) + sec->payload_len - 4;
-	if (!(max < r_buf_size (b))) {
+	if (!(ptr->offset.len = consume_init_expr_r (b, max, R_BIN_WASM_END_OF_CODE, NULL))) {
 		goto beach;
 	}
-	while (r_buf_tell (b) <= max) {
-		if (!(ptr = R_NEW0 (RBinWasmSymbol))) {
-			return ret;
-		}
-		ut32 tmp = 0;
-		read = consume_u32_r (b, max, &ptr->id);
-		consume_u32_r (b, max - read, &tmp);
-		if (tmp == R_BIN_WASM_STRING_LENGTH) {
-			tmp = R_BIN_WASM_STRING_LENGTH - 1;
-		}
-		ptr->name_len = tmp;
-		if (!(consume_str_r (b, max, tmp, ptr->name))) {
-			goto beach;
-		}
-		ptr->name[tmp] = 0;
-		if (!r_list_append (ret, ptr)) {
-			free (ptr);
-			// should it jump to beach?
-		}
-		ptr = NULL;
+	if (!(consume_u32_r (b, max, &ptr->size))) {
+		goto beach;
 	}
-	return ret;
+	ptr->data = r_buf_tell (b);
+	r_buf_seek (b, ptr->size, R_IO_SEEK_CUR);
+	return ptr;
+
 beach:
-	eprintf ("[wasm] error: beach symtab entries\n");
```

Comment:
Please keep `beach` errors.

Reply:
I kept them in a single place at https://github.com/radare/radare2/pull/13443/files#diff-bc8cad383202ffde95806a51d3a5c748R259 . Is that ok?

Reply:
ok